File: python/src/hanzo_zap/identity.py
"""
W3C Decentralized Identifier (DID) Implementation

Implements W3C DID Core 1.0 specification with support for:
- did:lux - Lux blockchain-anchored DIDs
- did:key - Self-certifying DIDs from cryptographic keys
- did:web - DNS-based DIDs

Example:
    >>> from hanzo_zap.identity import Did, DidMethod, NodeIdentity

    # Parse existing DID
    >>> did = parse_did("did:lux:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK")
    >>> did.method
    <DidMethod.LUX: 'lux'>

    # Create from ML-DSA public key
    >>> did = create_did_from_key(public_key_bytes)

    # Generate DID Document
    >>> doc = did.document()

    # Generate node identity
    >>> identity = generate_identity()
"""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional, Protocol, Tuple, Union

# Base58 alphabet (Bitcoin style)
BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

# Multibase prefix for base58btc
MULTIBASE_BASE58BTC = "z"

# Multicodec prefix for ML-DSA-65 public key (provisional)
MULTICODEC_MLDSA65 = bytes([0x13, 0x09])

# Expected ML-DSA-65 public key size
MLDSA_PUBLIC_KEY_SIZE = 1952


class IdentityError(Exception):
    """Identity-related error."""
    pass


def base58_encode(data: bytes) -> str:
    """Encode bytes to base58 (Bitcoin alphabet)."""
    num = int.from_bytes(data, "big")
    result = []
    while num > 0:
        num, remainder = divmod(num, 58)
        result.append(BASE58_ALPHABET[remainder])

    # Handle leading zeros
    for byte in data:
        if byte == 0:
            result.append(BASE58_ALPHABET[0])
        else:
            break

    return "".join(reversed(result))


def base58_decode(s: str) -> bytes:
    """Decode base58 string to bytes."""
    num = 0
    for char in s:
        num = num * 58 + BASE58_ALPHABET.index(char)

    # Calculate byte length
    result = []
    while num > 0:
        num, remainder = divmod(num, 256)
        result.append(remainder)

    # Handle leading ones (zeros in decoded)
    for char in s:
        if char == BASE58_ALPHABET[0]:
            result.append(0)
        else:
            break

    return bytes(reversed(result))


class DidMethod(Enum):
    """DID method identifier."""
    LUX = "lux"
    KEY = "key"
    WEB = "web"


class VerificationMethodType(Enum):
    """Verification method type."""
    JSON_WEB_KEY_2020 = "JsonWebKey2020"
    MULTIKEY = "Multikey"
    ML_DSA_65_VERIFICATION_KEY_2024 = "MlDsa65VerificationKey2024"


class ServiceType(Enum):
    """Service type."""
    ZAP_AGENT = "ZapAgent"
    DID_COMM_MESSAGING = "DIDCommMessaging"
    LINKED_DOMAINS = "LinkedDomains"
    CREDENTIAL_REGISTRY = "CredentialRegistry"


@dataclass
class ServiceEndpoint:
    """Service endpoint configuration."""
    uri: str
    accept: Optional[List[str]] = None
    routing_keys: Optional[List[str]] = None

    def to_dict(self) -> Union[str, Dict[str, Any]]:
        """Convert to JSON-serializable format."""
        if self.accept is None and self.routing_keys is None:
            return self.uri
        result: Dict[str, Any] = {"uri": self.uri}
        if self.accept:
            result["accept"] = self.accept
        if self.routing_keys:
            result["routingKeys"] = self.routing_keys
        return result


@dataclass
class VerificationMethod:
    """Verification method (public key) in DID Document."""
    id: str
    type: VerificationMethodType
    controller: str
    public_key_multibase: Optional[str] = None
    public_key_jwk: Optional[Dict[str, Any]] = None
    blockchain_account_id: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to JSON-serializable format."""
        result = {
            "id": self.id,
            "type": self.type.value,
            "controller": self.controller,
        }
        if self.public_key_multibase:
            result["publicKeyMultibase"] = self.public_key_multibase
        if self.public_key_jwk:
            result["publicKeyJwk"] = self.public_key_jwk
        if self.blockchain_account_id:
            result["blockchainAccountId"] = self.blockchain_account_id
        return result


@dataclass
class Service:
    """Service endpoint in DID Document."""
    id: str
    type: ServiceType
    service_endpoint: ServiceEndpoint

    def to_dict(self) -> Dict[str, Any]:
        """Convert to JSON-serializable format."""
        return {
            "id": self.id,
            "type": self.type.value,
            "serviceEndpoint": self.service_endpoint.to_dict(),
        }


@dataclass
class DidDocument:
    """W3C DID Document."""
    id: str
    context: List[str] = field(default_factory=lambda: [
        "https://www.w3.org/ns/did/v1",
        "https://w3id.org/security/suites/jws-2020/v1",
    ])
    controller: Optional[str] = None
    verification_method: List[VerificationMethod] = field(default_factory=list)
    authentication: List[str] = field(default_factory=list)
    assertion_method: List[str] = field(default_factory=list)
    key_agreement: List[str] = field(default_factory=list)
    capability_invocation: List[str] = field(default_factory=list)
    capability_delegation: List[str] = field(default_factory=list)
    service: List[Service] = field(default_factory=list)

    def primary_verification_method(self) -> Optional[VerificationMethod]:
        """Get the primary verification method."""
        return self.verification_method[0] if self.verification_method else None

    def get_verification_method(self, id: str) -> Optional[VerificationMethod]:
        """Get a verification method by ID."""
        for vm in self.verification_method:
            if vm.id == id:
                return vm
        return None

    def get_service(self, id: str) -> Optional[Service]:
        """Get a service by ID."""
        for svc in self.service:
            if svc.id == id:
                return svc
        return None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to JSON-serializable format."""
        result: Dict[str, Any] = {
            "@context": self.context,
            "id": self.id,
        }
        if self.controller:
            result["controller"] = self.controller
        if self.verification_method:
            result["verificationMethod"] = [vm.to_dict() for vm in self.verification_method]
        if self.authentication:
            result["authentication"] = self.authentication
        if self.assertion_method:
            result["assertionMethod"] = self.assertion_method
        if self.key_agreement:
            result["keyAgreement"] = self.key_agreement
        if self.capability_invocation:
            result["capabilityInvocation"] = self.capability_invocation
        if self.capability_delegation:
            result["capabilityDelegation"] = self.capability_delegation
        if self.service:
            result["service"] = [svc.to_dict() for svc in self.service]
        return result

    def to_json(self, indent: int = 2) -> str:
        """Serialize to JSON string."""
        return json.dumps(self.to_dict(), indent=indent)

    @classmethod
    def from_json(cls, json_str: str) -> "DidDocument":
        """Deserialize from JSON string."""
        data = json.loads(json_str)
        return cls.from_dict(data)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "DidDocument":
        """Create from dictionary."""
        verification_methods = []
        for vm_data in data.get("verificationMethod", []):
            verification_methods.append(VerificationMethod(
                id=vm_data["id"],
                type=VerificationMethodType(vm_data["type"]),
                controller=vm_data["controller"],
                public_key_multibase=vm_data.get("publicKeyMultibase"),
                public_key_jwk=vm_data.get("publicKeyJwk"),
                blockchain_account_id=vm_data.get("blockchainAccountId"),
            ))

        services = []
        for svc_data in data.get("service", []):
            endpoint_data = svc_data["serviceEndpoint"]
            if isinstance(endpoint_data, str):
                endpoint = ServiceEndpoint(uri=endpoint_data)
            else:
                endpoint = ServiceEndpoint(
                    uri=endpoint_data["uri"],
                    accept=endpoint_data.get("accept"),
                    routing_keys=endpoint_data.get("routingKeys"),
                )
            services.append(Service(
                id=svc_data["id"],
                type=ServiceType(svc_data["type"]),
                service_endpoint=endpoint,
            ))

        return cls(
            id=data["id"],
            context=data.get("@context", []),
            controller=data.get("controller"),
            verification_method=verification_methods,
            authentication=data.get("authentication", []),
            assertion_method=data.get("assertionMethod", []),
            key_agreement=data.get("keyAgreement", []),
            capability_invocation=data.get("capabilityInvocation", []),
            capability_delegation=data.get("capabilityDelegation", []),
            service=services,
        )


@dataclass
class Did:
    """W3C Decentralized Identifier (DID)."""
    method: DidMethod
    id: str

    def uri(self) -> str:
        """Get the full DID URI string."""
        return f"did:{self.method.value}:{self.id}"

    def __str__(self) -> str:
        return self.uri()

    def __hash__(self) -> int:
        return hash((self.method, self.id))

    def extract_key_material(self) -> bytes:
        """Extract raw key material from did:key or did:lux identifier."""
        if not self.id:
            raise IdentityError("empty DID identifier")

        if not self.id.startswith(MULTIBASE_BASE58BTC):
            raise IdentityError(
                f"unsupported multibase encoding: expected '{MULTIBASE_BASE58BTC}', "
                f"got '{self.id[0]}'"
            )

        # Decode base58btc (skip multibase prefix)
        try:
            decoded = base58_decode(self.id[1:])
        except Exception as e:
            raise IdentityError(f"invalid base58btc encoding: {e}")

        if len(decoded) < 2:
            raise IdentityError("DID identifier too short")

        # Skip multicodec prefix if it matches ML-DSA-65
        if decoded[:2] == MULTICODEC_MLDSA65:
            return decoded[2:]

        return decoded

    def document(self) -> DidDocument:
        """Generate a W3C DID Document for this DID."""
        did_uri = self.uri()

        # Create verification method based on DID type
        if self.method in (DidMethod.KEY, DidMethod.LUX):
            key_material = self.extract_key_material()
            blockchain_account_id = None
            if self.method == DidMethod.LUX:
                blockchain_account_id = f"lux:{key_material[:20].hex()}"

            verification_method = VerificationMethod(
                id=f"{did_uri}#keys-1",
                type=VerificationMethodType.JSON_WEB_KEY_2020,
                controller=did_uri,
                public_key_multibase=self.id,
                blockchain_account_id=blockchain_account_id,
            )
        else:
            verification_method = VerificationMethod(
                id=f"{did_uri}#keys-1",
                type=VerificationMethodType.JSON_WEB_KEY_2020,
                controller=did_uri,
            )

        # Create default service endpoint for ZAP protocol
        service = Service(
            id=f"{did_uri}#zap-agent",
            type=ServiceType.ZAP_AGENT,
            service_endpoint=ServiceEndpoint(uri=f"zap://{self.id}"),
        )

        return DidDocument(
            id=did_uri,
            verification_method=[verification_method],
            authentication=[f"{did_uri}#keys-1"],
            assertion_method=[f"{did_uri}#keys-1"],
            capability_invocation=[f"{did_uri}#keys-1"],
            service=[service],
        )


def parse_did(s: str) -> Did:
    """
    Parse a DID from a string in the format "did:method:id".

    Args:
        s: DID string to parse

    Returns:
        Parsed Did object

    Raises:
        IdentityError: If the DID string is invalid

    Example:
        >>> did = parse_did("did:lux:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK")
        >>> did.method
        <DidMethod.LUX: 'lux'>
    """
    if not s.startswith("did:"):
        raise IdentityError(f"invalid DID: must start with 'did:', got '{s}'")

    rest = s[4:]  # Skip "did:"
    parts = rest.split(":", 1)

    if len(parts) != 2:
        raise IdentityError(f"invalid DID format: expected 'did:method:id', got '{s}'")

    method_str, did_id = parts

    try:
        method = DidMethod(method_str)
    except ValueError:
        raise IdentityError(f"unknown DID method: {method_str}")

    if not did_id:
        raise IdentityError("DID identifier cannot be empty")

    return Did(method=method, id=did_id)


def create_did_from_key(public_key: bytes, method: DidMethod = DidMethod.KEY) -> Did:
    """
    Create a DID from an ML-DSA-65 public key.

    Args:
        public_key: ML-DSA-65 public key bytes (1952 bytes)
        method: DID method to use (KEY or LUX)

    Returns:
        New Did object

    Raises:
        IdentityError: If the public key is invalid

    Example:
        >>> did = create_did_from_key(public_key_bytes)
        >>> print(did)  # did:key:z6Mk...
    """
    if len(public_key) != MLDSA_PUBLIC_KEY_SIZE:
        raise IdentityError(
            f"invalid ML-DSA public key size: expected {MLDSA_PUBLIC_KEY_SIZE}, "
            f"got {len(public_key)}"
        )

    # Create multicodec-prefixed key
    prefixed = MULTICODEC_MLDSA65 + public_key

    # Encode with multibase (base58btc)
    encoded = base58_encode(prefixed)
    did_id = f"{MULTIBASE_BASE58BTC}{encoded}"

    return Did(method=method, id=did_id)


def create_did_from_web(domain: str, path: Optional[str] = None) -> Did:
    """
    Create a web DID from a domain and optional path.

    Args:
        domain: Domain name (e.g., "example.com")
        path: Optional path (e.g., "users/alice")

    Returns:
        New Did object with method=WEB

    Raises:
        IdentityError: If the domain is invalid

    Example:
        >>> did = create_did_from_web("example.com", "users/alice")
        >>> print(did)  # did:web:example.com:users:alice
    """
    if not domain:
        raise IdentityError("domain cannot be empty")

    if "/" in domain or ":" in domain:
        raise IdentityError(f"invalid domain for did:web: {domain}")

    if path:
        # Replace '/' with ':' per did:web spec
        path_parts = path.replace("/", ":")
        did_id = f"{domain}:{path_parts}"
    else:
        did_id = domain

    return Did(method=DidMethod.WEB, id=did_id)


class StakeRegistry(Protocol):
    """Protocol for stake registry implementations."""

    def get_stake(self, did: Did) -> int:
        """Get the staked amount for a DID."""
        ...

    def set_stake(self, did: Did, amount: int) -> None:
        """Set the staked amount for a DID."""
        ...

    def total_stake(self) -> int:
        """Get total staked amount across all DIDs."""
        ...


class InMemoryStakeRegistry:
    """In-memory stake registry for testing."""

    def __init__(self) -> None:
        self._stakes: Dict[str, int] = {}

    def get_stake(self, did: Did) -> int:
        """Get the staked amount for a DID."""
        return self._stakes.get(did.uri(), 0)

    def set_stake(self, did: Did, amount: int) -> None:
        """Set the staked amount for a DID."""
        self._stakes[did.uri()] = amount

    def total_stake(self) -> int:
        """Get total staked amount across all DIDs."""
        return sum(self._stakes.values())

    def has_sufficient_stake(self, did: Did, minimum: int) -> bool:
        """Check if a DID has sufficient stake."""
        return self.get_stake(did) >= minimum

    def stake_weight(self, did: Did) -> float:
        """Get the stake weight (0.0-1.0) for a DID relative to total."""
        stake = self.get_stake(did)
        total = self.total_stake()
        if total == 0:
            return 0.0
        return stake / total


@dataclass
class NodeIdentity:
    """
    Node identity combining DID with cryptographic keypair.

    Used for authenticated node participation in the ZAP network.
    """
    did: Did
    public_key: bytes
    stake: Optional[int] = None
    stake_registry: Optional[str] = None
    _signer: Optional[Any] = field(default=None, repr=False)

    def can_sign(self) -> bool:
        """Check if this node has signing capability."""
        return self._signer is not None

    def sign(self, message: bytes) -> bytes:
        """Sign a message with this node's private key."""
        if self._signer is None:
            raise IdentityError("no private key available for signing")
        return self._signer.sign(message)

    def verify(self, message: bytes, signature: bytes) -> bool:
        """Verify a signature against this node's public key."""
        try:
            from .crypto import PQSignature, PQ_AVAILABLE
            if not PQ_AVAILABLE:
                raise IdentityError("verification requires pqcrypto")

            if self._signer is not None:
                return self._signer.verify(message, signature)
            else:
                verifier = PQSignature.from_public_key(self.public_key)
                return verifier.verify(message, signature)
        except ImportError:
            raise IdentityError("verification requires pqcrypto")

    def document(self) -> DidDocument:
        """Get the DID document for this node identity."""
        return self.did.document()

    def with_stake(self, amount: int) -> "NodeIdentity":
        """Set the stake amount for this node."""
        self.stake = amount
        return self

    def with_registry(self, registry: str) -> "NodeIdentity":
        """Set the stake registry reference."""
        self.stake_registry = registry
        return self


def generate_identity(method: DidMethod = DidMethod.LUX) -> NodeIdentity:
    """
    Generate a new node identity with fresh ML-DSA-65 keypair.

    Args:
        method: DID method to use (default: LUX)

    Returns:
        New NodeIdentity with signing capability

    Raises:
        IdentityError: If pqcrypto is not available

    Example:
        >>> identity = generate_identity()
        >>> print(identity.did)  # did:lux:z6Mk...
        >>> identity.can_sign()
        True
    """
    try:
        from .crypto import PQSignature, PQ_AVAILABLE
        if not PQ_AVAILABLE:
            raise IdentityError("identity generation requires pqcrypto")

        signer = PQSignature.generate()
        public_key = signer.public_key
        did = create_did_from_key(public_key, method=method)

        return NodeIdentity(
            did=did,
            public_key=public_key,
            _signer=signer,
        )
    except ImportError:
        raise IdentityError("identity generation requires pqcrypto")


# Export public API
__all__ = [
    "IdentityError",
    "DidMethod",
    "VerificationMethodType",
    "ServiceType",
    "ServiceEndpoint",
    "VerificationMethod",
    "Service",
    "DidDocument",
    "Did",
    "parse_did",
    "create_did_from_key",
    "create_did_from_web",
    "StakeRegistry",
    "InMemoryStakeRegistry",
    "NodeIdentity",
    "generate_identity",
    "MLDSA_PUBLIC_KEY_SIZE",
]

File: typescript/src/identity.ts
/**
 * W3C Decentralized Identifier (DID) Implementation
 *
 * Implements W3C DID Core 1.0 specification with support for:
 * - did:lux - Lux blockchain-anchored DIDs
 * - did:key - Self-certifying DIDs from cryptographic keys
 * - did:web - DNS-based DIDs
 *
 * @example
 * ```typescript
 * import { Did, DidMethod, NodeIdentity, parseDid, createDidFromKey } from './identity';
 *
 * // Parse existing DID
 * const did = parseDid("did:lux:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK");
 *
 * // Create from ML-DSA public key
 * const did = createDidFromKey(publicKeyBytes);
 *
 * // Generate DID Document
 * const doc = did.document();
 *
 * // Generate node identity
 * const identity = await generateIdentity();
 * ```
 */

// Base58 alphabet (Bitcoin style)
const BASE58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

// Multibase prefix for base58btc
const MULTIBASE_BASE58BTC = 'z';

// Multicodec prefix for ML-DSA-65 public key (provisional)
const MULTICODEC_MLDSA65 = new Uint8Array([0x13, 0x09]);

// Expected ML-DSA-65 public key size
export const MLDSA_PUBLIC_KEY_SIZE = 1952;

/**
 * Identity-related error
 */
export class IdentityError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'IdentityError';
  }
}

/**
 * Encode bytes to base58 (Bitcoin alphabet)
 */
function base58Encode(data: Uint8Array): string {
  let num = BigInt(0);
  for (let i = 0; i < data.length; i++) {
    num = num * BigInt(256) + BigInt(data[i]!);
  }

  const result: string[] = [];
  while (num > BigInt(0)) {
    const remainder = Number(num % BigInt(58));
    num = num / BigInt(58);
    result.push(BASE58_ALPHABET[remainder]!);
  }

  // Handle leading zeros
  for (let i = 0; i < data.length; i++) {
    if (data[i] === 0) {
      result.push(BASE58_ALPHABET[0]!);
    } else {
      break;
    }
  }

  return result.reverse().join('');
}

/**
 * Decode base58 string to bytes
 */
function base58Decode(s: string): Uint8Array {
  let num = BigInt(0);
  for (const char of s) {
    const index = BASE58_ALPHABET.indexOf(char);
    if (index === -1) {
      throw new IdentityError(`invalid base58 character: ${char}`);
    }
    num = num * BigInt(58) + BigInt(index);
  }

  // Convert to bytes
  const bytes: number[] = [];
  while (num > BigInt(0)) {
    bytes.push(Number(num % BigInt(256)));
    num = num / BigInt(256);
  }

  // Handle leading ones (zeros in decoded)
  for (const char of s) {
    if (char === BASE58_ALPHABET[0]) {
      bytes.push(0);
    } else {
      break;
    }
  }

  return new Uint8Array(bytes.reverse());
}

/**
 * DID method identifier
 */
export enum DidMethod {
  Lux = 'lux',
  Key = 'key',
  Web = 'web',
}

/**
 * Verification method type
 */
export enum VerificationMethodType {
  JsonWebKey2020 = 'JsonWebKey2020',
  Multikey = 'Multikey',
  MlDsa65VerificationKey2024 = 'MlDsa65VerificationKey2024',
}

/**
 * Service type
 */
export enum ServiceType {
  ZapAgent = 'ZapAgent',
  DIDCommMessaging = 'DIDCommMessaging',
  LinkedDomains = 'LinkedDomains',
  CredentialRegistry = 'CredentialRegistry',
}

/**
 * Service endpoint configuration
 */
export interface ServiceEndpoint {
  uri: string;
  accept?: string[];
  routingKeys?: string[];
}

/**
 * Verification method (public key) in DID Document
 */
export interface VerificationMethod {
  id: string;
  type: VerificationMethodType;
  controller: string;
  publicKeyMultibase?: string;
  publicKeyJwk?: Record<string, unknown>;
  blockchainAccountId?: string;
}

/**
 * Service endpoint in DID Document
 */
export interface Service {
  id: string;
  type: ServiceType;
  serviceEndpoint: string | ServiceEndpoint;
}

/**
 * W3C DID Document
 */
export interface DidDocument {
  '@context': string[];
  id: string;
  controller?: string;
  verificationMethod?: VerificationMethod[];
  authentication?: string[];
  assertionMethod?: string[];
  keyAgreement?: string[];
  capabilityInvocation?: string[];
  capabilityDelegation?: string[];
  service?: Service[];
}

/**
 * W3C Decentralized Identifier (DID)
 */
export interface Did {
  method: DidMethod;
  id: string;
}

/**
 * Create a DID URI string
 */
export function didUri(did: Did): string {
  return `did:${did.method}:${did.id}`;
}

/**
 * Extract raw key material from did:key or did:lux identifier
 */
export function extractKeyMaterial(did: Did): Uint8Array {
  if (!did.id) {
    throw new IdentityError('empty DID identifier');
  }

  if (!did.id.startsWith(MULTIBASE_BASE58BTC)) {
    throw new IdentityError(
      `unsupported multibase encoding: expected '${MULTIBASE_BASE58BTC}', got '${did.id[0]}'`
    );
  }

  // Decode base58btc (skip multibase prefix)
  const decoded = base58Decode(did.id.slice(1));

  if (decoded.length < 2) {
    throw new IdentityError('DID identifier too short');
  }

  // Skip multicodec prefix if it matches ML-DSA-65
  if (decoded[0] === MULTICODEC_MLDSA65[0] && decoded[1] === MULTICODEC_MLDSA65[1]) {
    return decoded.slice(2);
  }

  return decoded;
}

/**
 * Generate a W3C DID Document for a DID
 */
export function generateDocument(did: Did): DidDocument {
  const uri = didUri(did);

  let verificationMethod: VerificationMethod;
  if (did.method === DidMethod.Key || did.method === DidMethod.Lux) {
    const keyMaterial = extractKeyMaterial(did);

    if (did.method === DidMethod.Lux) {
      // Create blockchain account ID from first 20 bytes
      const accountBytes = keyMaterial.slice(0, 20);
      const blockchainAccountId = `lux:${Array.from(accountBytes)
        .map((b) => b.toString(16).padStart(2, '0'))
        .join('')}`;

      verificationMethod = {
        id: `${uri}#keys-1`,
        type: VerificationMethodType.JsonWebKey2020,
        controller: uri,
        publicKeyMultibase: did.id,
        blockchainAccountId,
      };
    } else {
      verificationMethod = {
        id: `${uri}#keys-1`,
        type: VerificationMethodType.JsonWebKey2020,
        controller: uri,
        publicKeyMultibase: did.id,
      };
    }
  } else {
    verificationMethod = {
      id: `${uri}#keys-1`,
      type: VerificationMethodType.JsonWebKey2020,
      controller: uri,
    };
  }

  const service: Service = {
    id: `${uri}#zap-agent`,
    type: ServiceType.ZapAgent,
    serviceEndpoint: `zap://${did.id}`,
  };

  return {
    '@context': [
      'https://www.w3.org/ns/did/v1',
      'https://w3id.org/security/suites/jws-2020/v1',
    ],
    id: uri,
    verificationMethod: [verificationMethod],
    authentication: [`${uri}#keys-1`],
    assertionMethod: [`${uri}#keys-1`],
    capabilityInvocation: [`${uri}#keys-1`],
    service: [service],
  };
}

/**
 * Parse a DID from a string in the format "did:method:id"
 *
 * @param s - DID string to parse
 * @returns Parsed Did object
 * @throws IdentityError if the DID string is invalid
 *
 * @example
 * ```typescript
 * const did = parseDid("did:lux:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK");
 * console.log(did.method); // DidMethod.Lux
 * ```
 */
export function parseDid(s: string): Did {
  if (!s.startsWith('did:')) {
    throw new IdentityError(`invalid DID: must start with 'did:', got '${s}'`);
  }

  const rest = s.slice(4); // Skip "did:"
  const colonIndex = rest.indexOf(':');

  if (colonIndex === -1) {
    throw new IdentityError(`invalid DID format: expected 'did:method:id', got '${s}'`);
  }

  const methodStr = rest.slice(0, colonIndex);
  const didId = rest.slice(colonIndex + 1);

  let method: DidMethod;
  switch (methodStr) {
    case 'lux':
      method = DidMethod.Lux;
      break;
    case 'key':
      method = DidMethod.Key;
      break;
    case 'web':
      method = DidMethod.Web;
      break;
    default:
      throw new IdentityError(`unknown DID method: ${methodStr}`);
  }

  if (!didId) {
    throw new IdentityError('DID identifier cannot be empty');
  }

  return { method, id: didId };
}

/**
 * Create a DID from an ML-DSA-65 public key
 *
 * @param publicKey - ML-DSA-65 public key bytes (1952 bytes)
 * @param method - DID method to use (KEY or LUX)
 * @returns New Did object
 * @throws IdentityError if the public key is invalid
 *
 * @example
 * ```typescript
 * const did = createDidFromKey(publicKeyBytes);
 * console.log(didUri(did)); // did:key:z6Mk...
 * ```
 */
export function createDidFromKey(
  publicKey: Uint8Array,
  method: DidMethod = DidMethod.Key
): Did {
  if (publicKey.length !== MLDSA_PUBLIC_KEY_SIZE) {
    throw new IdentityError(
      `invalid ML-DSA public key size: expected ${MLDSA_PUBLIC_KEY_SIZE}, got ${publicKey.length}`
    );
  }

  // Create multicodec-prefixed key
  const prefixed = new Uint8Array(MULTICODEC_MLDSA65.length + publicKey.length);
  prefixed.set(MULTICODEC_MLDSA65, 0);
  prefixed.set(publicKey, MULTICODEC_MLDSA65.length);

  // Encode with multibase (base58btc)
  const encoded = base58Encode(prefixed);
  const id = `${MULTIBASE_BASE58BTC}${encoded}`;

  return { method, id };
}

/**
 * Create a web DID from a domain and optional path
 *
 * @param domain - Domain name (e.g., "example.com")
 * @param path - Optional path (e.g., "users/alice")
 * @returns New Did object with method=WEB
 * @throws IdentityError if the domain is invalid
 *
 * @example
 * ```typescript
 * const did = createDidFromWeb("example.com", "users/alice");
 * console.log(didUri(did)); // did:web:example.com:users:alice
 * ```
 */
export function createDidFromWeb(domain: string, path?: string): Did {
  if (!domain) {
    throw new IdentityError('domain cannot be empty');
  }

  if (domain.includes('/') || domain.includes(':')) {
    throw new IdentityError(`invalid domain for did:web: ${domain}`);
  }

  let id: string;
  if (path) {
    // Replace '/' with ':' per did:web spec
    const pathParts = path.replace(/\//g, ':');
    id = `${domain}:${pathParts}`;
  } else {
    id = domain;
  }

  return { method: DidMethod.Web, id };
}

/**
 * Stake registry interface
 */
export interface StakeRegistry {
  getStake(did: Did): Promise<bigint>;
  setStake(did: Did, amount: bigint): Promise<void>;
  totalStake(): Promise<bigint>;
  hasSufficientStake(did: Did, minimum: bigint): Promise<boolean>;
  stakeWeight(did: Did): Promise<number>;
}

/**
 * In-memory stake registry for testing
 */
export class InMemoryStakeRegistry implements StakeRegistry {
  private stakes: Map<string, bigint> = new Map();

  async getStake(did: Did): Promise<bigint> {
    return this.stakes.get(didUri(did)) ?? BigInt(0);
  }

  async setStake(did: Did, amount: bigint): Promise<void> {
    this.stakes.set(didUri(did), amount);
  }

  async totalStake(): Promise<bigint> {
    let total = BigInt(0);
    this.stakes.forEach((stake) => {
      total += stake;
    });
    return total;
  }

  async hasSufficientStake(did: Did, minimum: bigint): Promise<boolean> {
    const stake = await this.getStake(did);
    return stake >= minimum;
  }

  async stakeWeight(did: Did): Promise<number> {
    const stake = await this.getStake(did);
    const total = await this.totalStake();
    if (total === BigInt(0)) {
      return 0.0;
    }
    return Number(stake) / Number(total);
  }
}

/**
 * Signer interface for cryptographic operations
 */
export interface Signer {
  sign(message: Uint8Array): Promise<Uint8Array>;
  verify(message: Uint8Array, signature: Uint8Array): Promise<boolean>;
  publicKey: Uint8Array;
}

/**
 * Node identity combining DID with cryptographic keypair
 *
 * Used for authenticated node participation in the ZAP network.
 */
export class NodeIdentity {
  public readonly did: Did;
  public readonly publicKey: Uint8Array;
  public stake: bigint | undefined;
  public stakeRegistry: string | undefined;
  private signer: Signer | undefined;

  constructor(did: Did, publicKey: Uint8Array, signer?: Signer) {
    this.did = did;
    this.publicKey = publicKey;
    this.signer = signer;
  }

  /**
   * Check if this node has signing capability
   */
  canSign(): boolean {
    return this.signer !== undefined;
  }

  /**
   * Sign a message with this node's private key
   */
  async sign(message: Uint8Array): Promise<Uint8Array> {
    if (!this.signer) {
      throw new IdentityError('no private key available for signing');
    }
    return this.signer.sign(message);
  }

  /**
   * Verify a signature against this node's public key
   */
  async verify(message: Uint8Array, signature: Uint8Array): Promise<boolean> {
    if (this.signer) {
      return this.signer.verify(message, signature);
    }
    throw new IdentityError('verification requires a signer implementation');
  }

  /**
   * Get the DID document for this node identity
   */
  document(): DidDocument {
    return generateDocument(this.did);
  }

  /**
   * Get the DID URI string
   */
  uri(): string {
    return didUri(this.did);
  }

  /**
   * Set the stake amount for this node
   */
  withStake(amount: bigint): this {
    this.stake = amount;
    return this;
  }

  /**
   * Set the stake registry reference
   */
  withRegistry(registry: string): this {
    this.stakeRegistry = registry;
    return this;
  }
}

/**
 * Generate a new node identity with fresh ML-DSA-65 keypair
 *
 * Note: This requires a Signer implementation to be provided.
 * In production, use a proper cryptographic library.
 *
 * @param signer - Signer implementation with ML-DSA-65 keypair
 * @param method - DID method to use (default: LUX)
 * @returns New NodeIdentity with signing capability
 *
 * @example
 * ```typescript
 * // With a signer implementation
 * const identity = generateIdentity(signer);
 * console.log(identity.uri()); // did:lux:z6Mk...
 * identity.canSign(); // true
 * ```
 */
export function generateIdentity(
  signer: Signer,
  method: DidMethod = DidMethod.Lux
): NodeIdentity {
  const publicKey = signer.publicKey;

  if (publicKey.length !== MLDSA_PUBLIC_KEY_SIZE) {
    throw new IdentityError(
      `invalid ML-DSA public key size: expected ${MLDSA_PUBLIC_KEY_SIZE}, got ${publicKey.length}`
    );
  }

  const did = createDidFromKey(publicKey, method);
  return new NodeIdentity(did, publicKey, signer);
}

/**
 * Create a NodeIdentity from an existing DID and public key (verification only)
 *
 * @param did - Existing DID
 * @param publicKey - Public key bytes
 * @returns NodeIdentity without signing capability
 */
export function createNodeIdentity(did: Did, publicKey: Uint8Array): NodeIdentity {
  return new NodeIdentity(did, publicKey);
}
